app/config/AWS.scala

```diff
@@ -3,52 +3,69 @@ package config
-    new InstanceProfileCredentialsProvider(),
+    InstanceProfileCredentialsProvider.getInstance(),
     new ProfileCredentialsProvider("workflow")
   )
   lazy val composerAwsCredentialsProvider = new AWSCredentialsProviderChain(
     new EnvironmentVariableCredentialsProvider(),
     new SystemPropertiesCredentialsProvider(),
-    new InstanceProfileCredentialsProvider(),
+    InstanceProfileCredentialsProvider.getInstance(),
     new ProfileCredentialsProvider("composer")
   )
 
-  def eC2Client = region.createClient(classOf[AmazonEC2Client], workflowAwsCredentialsProvider, null)
-  def s3Client = region.createClient(classOf[AmazonS3Client], composerAwsCredentialsProvider, null)
-  lazy val sesClient = region.createClient(classOf[AmazonSimpleEmailServiceClient],
-    composerAwsCredentialsProvider, null)
-
-  def dynamoDbClient = region.createClient(classOf[AmazonDynamoDBClient], composerAwsCredentialsProvider, null)
-
-  def getInstanceId = Option(EC2MetadataUtils.getInstanceId)
-
-  def readTag(tagName: String, instanceId: Option[String])(implicit ec2Client: AmazonEC2Client): Option[String] = {
-    instanceId.flatMap { id =>
-      val tagsResult = ec2Client.describeTags(
-        new DescribeTagsRequest().withFilters(
-          new Filter("resource-type").withValues("instance"),
-          new Filter("resource-id").withValues(id),
-          new Filter("key").withValues(tagName)
-        )
-      )
-      tagsResult.getTags.asScala
-        .find(_.getKey == tagName)
-        .map(_.getValue)
+  lazy val asgClient = AmazonAutoScalingClientBuilder.standard().withRegion(region).withCredentials(composerAwsCredentialsProvider).build()
+  lazy val ssmClient = AWSSimpleSystemsManagementClientBuilder.standard().withRegion(region).withCredentials(composerAwsCredentialsProvider).build()
+
+  lazy val s3Client = AmazonS3ClientBuilder.standard().withRegion(region).withCredentials(composerAwsCredentialsProvider).build()
+  lazy val sesClient = AmazonSimpleEmailServiceClientBuilder.standard().withRegion(region).withCredentials(composerAwsCredentialsProvider).build()
+  lazy val dynamoDbClient = AmazonDynamoDBClientBuilder.standard().withRegion(region).withCredentials(composerAwsCredentialsProvider).build()
+
+  def readTags(): Option[InstanceTags] = {
+    // We read tags from the AutoScalingGroup rather than the instance itself to avoid problems where the
+    // tags have not been applied to the instance before we start up (they are eventually consistent)
+    for {
+      instanceId <- Option(EC2MetadataUtils.getInstanceId)
+      asg <- getAutoscalingGroupName(instanceId)
+      tags <- getTags(asg)
+
+      stack <- tags.get("Stack")
+      app <- tags.get("App")
+      stage <- tags.get("Stage")
+    } yield {
+      InstanceTags(stack, app, stage)
     }
   }
+
+  private def getAutoscalingGroupName(instanceId: String): Option[String] = {
+    val request = new DescribeAutoScalingInstancesRequest().withInstanceIds(instanceId)
```

Comment:
does the policy in cfn need to be updated to allow this?

Reply:
It's already `ec2:Describe*` which covers `DescribeTags` (used before) and `DescribeAutoScalingInstances`. https://github.com/guardian/editorial-tools-platform/blob/master/cloudformation/login.gutools/login-tool.yaml#L99.

We could raise a PR to restrict it down to just `DescribeAutoScalingInstances`.